### PR TITLE
[AMD] [Fix] Fix --attention-backend triton work for DeepSeek MLA on MI355 (null-K + decode dispatch + RoPE)

### PR DESCRIPTION
**Pull request by yichiche**
## Motivation

DeepSeek MLA models could not run with `--attention-backend triton` on gfx95 (MI300/MI355) when `SGLANG_USE_AITER=1` (the ROCm image default). Triton is the common NVIDIA/AMD path and should not GPU-fault or produce wrong output on a valid backend selection.

Root cause is a recurring anti-pattern: several gfx95/aiter fused MLA paths are gated on env vars (`SGLANG_USE_AITER`, `SGLANG_ROCM_FUSED_DECODE_MLA`) or on "not in the absorb-core backend list", **not on the actually selected attention backend**. As a result the triton backend inherits aiter-only fused paths that it cannot consume:

1. **Prefill null-K GPU fault** — in `forward_absorb_core`, triton fell into the aiter-fused `fused_qk_rope_cat_and_cache_mla` path, which writes K into the KV cache and returns an empty `k` (numel=0, null ptr) for the aiter kernel to read from the buffer. Triton's extend kernel reads `k` directly → `Memory access fault on address (nil)`.
2. **Decode dispatch fault** — `_dispatch_mla_subtype` honored `SGLANG_ROCM_FUSED_DECODE_MLA=1` for any backend, routing triton decode to `MLA_FUSED_ROPE_ROCM` (`forward_absorb_fused_mla_rope_prepare`), which is neither MXFP4-aware (`bmm` sees packed `w_kc`, 64 vs 128) nor triton-metadata-compatible (`ForwardMetadata` unpack).
3. **Accuracy 0.03** — `_skip_rope_for_aiter_fused_mla` returned True for triton (backend not in the absorb-core list), so RoPE was silently skipped in prepare while triton took the standard cat path → garbage output.

## Modifications

`forward_mla.py`:
- `forward_absorb_core`: gate the aiter-fused path to `self.current_attention_backend == "aiter"` so triton builds a real `k = torch.cat([k_nope, k_pe])` with `save_kv_cache=True`.
- rope-in-prepare condition: allow triton through (`or self.current_attention_backend == "triton"`).
- `_skip_rope_for_aiter_fused_mla`: restrict to `== "aiter"` so triton applies RoPE in prepare (the 0.03 → 0.945 fix).

`attention_backend_handler.py`:
- `_dispatch_mla_subtype`: gate `MLA_FUSED_ROPE_ROCM` on `current_attention_backend == "aiter"` so triton (and other backends) decode via the standard `MLA` path.

All changes are backend-scoped; aiter/fa3/flashinfer/CUDA paths are unchanged.

## Accuracy Tests

DeepSeek-R1-MXFP4 (Quark MXFP4), MI355X gfx950, tp4, `--kv-cache-dtype fp8_e4m3`, gsm8k (200q):

| Backend | Config | gsm8k |
|---|---|---|
| aiter (baseline) | + spec + cuda graph | 0.955 |
| **triton (this PR)** | **no spec** | **0.945** |

`--attention-backend triton` now runs DeepSeek-R1-MXFP4 end-to-end with correct output.

## Scope / Known Limitation

This PR makes triton MLA correct for **non-speculative** decode. `--attention-backend triton` **+ EAGLE speculative decoding is still broken** (GPU fault in the NextN draft/verify path, independent of cuda graph) and is out of scope here — that is a separate draft-path bug. For DeepSeek MLA with speculative decoding, continue using `--attention-backend aiter` (fully working, 0.955).

## Checklist

- [x] Backend-scoped; no change to aiter/CUDA behavior.
- [x] Verified triton correctness on real DeepSeek-R1-MXFP4 (0.945).
- [x] Follows `[Tag] Description` commit/PR convention.

---
### CI States

Latest PR Test (Base): :x: [Run #29228875042](https://github.com/sgl-project/sglang/actions/runs/29228875042)
Latest PR Test (Extra): :x: [Run #29228874957](https://github.com/sgl-project/sglang/actions/runs/29228874957)

### CODE REVIEW

**Comment by yichiche** [comment]
@amd-bot ci-status

**Comment by yichiche** [comment]
- To close the gap in PR CI (not nightly), I added test/registered/models/test_deepseek_mla_dispatch.py — a hermetic unit test (no GPU/model) registered on the CUDA base-b + AMD stage-b-test-1-gpu-small-amd-mi35x lanes. It pins _dispatch_mla_subtype so the triton+decode+rocm_fused_decode_mla case stays on plain MLA instead of the aiter-only
   MLA_FUSED_ROPE_ROCM fast path (the one that GPU-faults on gfx95). Proven guard: reverting the == "aiter" gate makes test_hip_triton_decode_stays_plain_mla fail; with the fix all 3 cases pass.
- Full e2e (R1-MXFP4 + NextN + --attention-backend triton, nightly recipe) was launched and got through arg-parse + weight-load before hitting an unrelated OOM from other jobs on the shared box; recipe is valid and can be re-run when ≥4 GPUs are free.

```
python -m pytest \
    ../test/registered/models/test_deepseek_mla_dispatch.py -v```
    
../test/registered/models/test_deepseek_mla_dispatch.py::TestDispatchMLASubtype::test_hip_aiter_decode_takes_fused_rope PASSED                                                     [ 50%]
../test/registered/models/test_deepseek_mla_dispatch.py::TestDispatchMLASubtype::test_hip_aiter_extend_stays_plain_mla PASSED                                                      [ 75%]
../test/registered/models/test_deepseek_mla_dispatch.py::TestDispatchMLASubtype::test_hip_triton_decode_stays_plain_mla PASSED                                                     [100%]
```

All pass.

File: test/registered/unit/models/test_deepseek_mla_dispatch.py
Comment (hnyls2002) [inline test/registered/unit/models/test_deepseek_mla_dispatch.py:1]: Put this under the unit test folder under register rather than `models/`; this folder would be deprecated

**Review by hnyls2002 — APPROVED** [review]

File: test/registered/unit/models/test_deepseek_mla_dispatch.py
Comment (HaiShaw) [reply test/registered/unit/models/test_deepseek_mla_dispatch.py:1]: @yichiche Yes, move to a place under `test/registered/unit`, then rerun ci

Comment (yichiche) [reply test/registered/unit/models/test_deepseek_mla_dispatch.py:1]: Moved the file to test/registered/unit, CI kicked off.

**Review by HaiShaw — APPROVED** [review]

**Comment by yichiche** [comment]
@HaiShaw CI completed, ready to merge.
